File: code/php/sdk/src/events/MergeSignals.php
<?php
/**
 * @copyright Copyright (c) PutYourLightsOn
 */

namespace starfederation\datastar\events;

use starfederation\datastar\Consts;
use starfederation\datastar\enums\EventType;

class MergeSignals implements EventInterface
{
    use EventTrait;

    public string $data;
    public bool $onlyIfMissing = Consts::DEFAULT_MERGE_SIGNALS_ONLY_IF_MISSING;

    public function __construct(string $data, array $options = [])
    {
        $this->data = $data;

        foreach ($options as $key => $value) {
            $this->$key = $value;
        }
    }

    /**
     * @inerhitdoc
     */
    public function getEventType(): EventType
    {
        return EventType::MergeSignals;
    }

    /**
     * @inerhitdoc
     */
    public function getDataLines(): array
    {
        $dataLines = [];

        if ($this->onlyIfMissing !== Consts::DEFAULT_MERGE_SIGNALS_ONLY_IF_MISSING) {
            $dataLines[] = $this->getDataLine(Consts::ONLY_IF_MISSING_DATALINE_LITERAL, $this->getBooleanAsString($this->onlyIfMissing));
        }

        return array_merge(
            $dataLines,
            $this->getMultiDataLines(Consts::SIGNALS_DATALINE_LITERAL, $this->data),
        );
    }
}

File: code/java/src/main/java/sdk/events/mergesignals/MergeSignals.java
package sdk.events.mergesignals;

import sdk.enums.EventType;
import sdk.events.EventInterface;
import sdk.events.SSEOptions;

import java.util.*;

import static sdk.Consts.*;

public class MergeSignals implements EventInterface {

    private final String data;
    private final SSEOptions sseOptions;
    private final MergeSignalsOptions mergeSignalsOptions;

    public MergeSignals(String data, SSEOptions sseOptions, MergeSignalsOptions mergeSignalsOptions) {
        if (data == null || data.isBlank()) {
            throw new IllegalArgumentException("Data cannot be null or empty.");
        }
        this.data = data.trim();
        this.sseOptions = Objects.requireNonNullElse(sseOptions, SSEOptions.create());
        this.mergeSignalsOptions = Objects.requireNonNullElse(mergeSignalsOptions, MergeSignalsOptions.create());
    }

    @Override
    public EventType getEventType() {
        return EventType.MERGE_SIGNALS;
    }

    @Override
    public Map<String, Object> getOptions() {
        Map<String, Object> options = new HashMap<>();
        if (sseOptions.getEventId() != null) {
            options.put(EVENT_ID_DATALINE_LITERAL, sseOptions.getEventId());
        }
        if (sseOptions.getRetryDuration() != DEFAULT_SSE_RETRY_DURATION) {
            options.put(RETRY_DURATION_DATALINE_LITERAL, sseOptions.getRetryDuration());
        }
        return options;
    }

    @Override
    public String[] getDataLines() {
        List<String> dataLines = new ArrayList<>();

        // Add onlyIfMissing if true
        if (mergeSignalsOptions.isOnlyIfMissing()) {
            dataLines.add(getDataLine(ONLY_IF_MISSING_DATALINE_LITERAL, getBooleanAsString(true)));
        }

        // Add data content line by line, prefixed with "signals"
        Arrays.stream(data.split("\n"))
                .map(String::trim)
                .filter(line -> !line.isEmpty())
                .forEach(line -> dataLines.add(getDataLine(SIGNALS_DATALINE_LITERAL, line)));

        return dataLines.toArray(String[]::new);
    }

    @Override
    public String getBooleanAsString(boolean value) {
        return value ? "true" : "false";
    }

    @Override
    public String getDataLine(String key, String value) {
        return "data: " + key + " " + value;
    }
}
